iothub/device/src/Pipeline/RetryDelegatingHandler.cs

```diff
@@ -699,6 +663,55 @@ private async Task EnsureOpenedAsync(bool withRetry, CancellationToken cancellat
             }
         }
 
+        public override async Task CloseAsync(CancellationToken cancellationToken)
+        {
+            await _handlerSemaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+            try
+            {
+                if (!_openCalled)
+                {
+                    return;
+                }
+
+                if (Logging.IsEnabled)
+                    Logging.Enter(this, cancellationToken, nameof(CloseAsync));
+
+                _handleDisconnectCts.Cancel();
+                await base.CloseAsync(cancellationToken).ConfigureAwait(false);
+            }
+            finally
+            {
+                if (Logging.IsEnabled)
+                    Logging.Exit(this, cancellationToken, nameof(CloseAsync));
+
+                _handlerSemaphore?.Release();
+                Dispose(true);
+            }
+        }
+
+        private async Task EnsureOpenedAsync()
```

Comment:
Why not pass in the cancellation token? If the user wants to cancel, this should cancel too.

Reply:
Added requested changes.